PPG full implementation

breakds opened the pull request:
# Motivation

This is to implement #965. 

![ppg](https://user-images.githubusercontent.com/1111035/137959366-097791a7-ad34-48a2-88ce-259305cdb38d.png)


# Review Guide

Various depended changes has been split to other PRs and submitted. Even though this final PR is big.

Please be advised that the main changes are

1. `alf/algorithms/ppg/ppg_aux_phase_loss.py` - This adds the auxiliary phase loss implementation. PPG has two phases (See graph above) where the main phase use the PPO Loss, the auxiliary phase needs its own loss implementation, and this file implements that.
2. `alf/algorithms/ppo_loss.py` - Added an option to compute the advantages inside the loss. This is actually the default behavior of actor critic loss.
3. `alf/algorithms/ppg_algorithm.py` - The main PPG algorithm implementation. It is specifically designed to have a context manager that can switch the mode to "Auxiliary Phase". Auxliliary phase update is called periodically in `after_train_iter`, and when activated, certain parameters such as "which replay buffer to use" is temporarily modified until the auxliliary update is done.
4. There are two sample configurations for PPG algorithm added: `ppg_procgen_bossfight_conf.py` and `ppg_cart_pole`.

# Testing

Mainly on procgen bossfight environment, where the perfect score is `15`.

![ppg_bossfight_1](https://user-images.githubusercontent.com/1111035/137960524-4d8905d4-e7b3-4d80-bbb2-ec2193492e61.png)

Both curve above comes from running the PPG algorithm, and the difference is

1. The orange curve is PPG algorithm with aux phase update every 32 iteration
2. The green curve is PPG algorithm without any aux phase update, meaning that it is **effectively just PPO**.

We can see that 

- PPG reaches a relatively high score pretty quickly
- At the same number of (main) training iterations, PPG has a higher reward.
- PPO reaches 9 in the end (6000 iterations), which took it 15 hours. PPG reaches 9 in 780 iterations, which took 3 hours.

This roughly resembles the experiment in the paper (though my settings is slightly different from the paper as I am training on 1 card while the paper is on 4 cards)

![ppg_paper_bossfight](https://user-images.githubusercontent.com/1111035/137961555-96cbc28d-02cb-426f-a88c-544f7e4df8b2.png)

breakds committed:
Implement the PPG Algorithm (#965)

breakds commented:
This PR is now ready for review. Thanks a lot!

emailweixu commented on alf/algorithms/ppg_algorithm.py:
```diff
@@ -0,0 +1,468 @@
+       auxiliary phase losses (different from PPO loss, see
+       algorithms/ppg/ppg_aux_phase_loss.py for details). Auxiliary phase
+       updates does not require new rollouts. Instead it is performed on all of
+       the experience collected since the last auxiliary phase update.
```
add a reference to ppg paper

breakds replied:
Reference added.

emailweixu commented on alf/algorithms/ppg_algorithm.py:
```diff
@@ -0,0 +1,468 @@
+
+    @property
+    def on_policy(self) -> bool:
+        return False
```
this is already implemented in OffPolicyAlgorithm

breakds replied:
Removed this.

emailweixu commented on alf/algorithms/ppg_algorithm.py:
```diff
@@ -0,0 +1,468 @@
+            config (TrainerConfig): config for training. config only needs to be
+                provided to the algorithm which performs ``train_iter()`` by
+                itself.
+            encoding_network_ctor (Callable[..., Network]): Function to
```
Comment how is this function called. And I think kernel_initializer should not be its required argument since it can always be configured if necessary.

breakds replied:
Removed `kernel_intializer` from the code, and updated the argument doc string to explicit state how this function should be called.

emailweixu commented on alf/algorithms/ppg_algorithm.py:
```diff
@@ -0,0 +1,468 @@
+        # specific to the auxiliary update phase.
+        self._aux_phase = PPGPhaseContext(
+            ppg_algorithm=self,
+            network=dual_actor_value_network.copy(),
```
It's not obvious why copy() is needed for everyone. Perhaps add some comment. (i.e. In ALF, one parameter cannot not be handled by multiple optimizers)

breakds replied:
Documentation added.

emailweixu commented on alf/algorithms/ppg_algorithm.py:
```diff
@@ -0,0 +1,468 @@
+                max_length=max_length,
+                prioritized_sampling=False,
+                num_earliest_frames_ignored=self._num_earliest_frames_ignored,
+                name=f'aux_replay_buffer')
```
no need to have 'f'

emailweixu commented on alf/algorithms/ppg_algorithm.py:
```diff
@@ -0,0 +1,468 @@
+
+        """
+        policy_step = self._policy_phase.network_forward(inputs, state)
+        self._observe_for_aux_replay(inputs.untransformed.cpu(), state,
```
To really integrate with Agent, it should store `inputs` instead of untransformed to the replay buffer. This is because Agent may change the original time step (e.g., by representation learner, or goal generator) before passing it to RL algorithm.

breakds replied:
This is previously required but not needed now after #1036. Removed it.

I was not aware that agent may change the original time step. Thanks for mentioning that!

emailweixu commented on alf/algorithms/ppo_loss.py:
```diff
@@ -31,6 +31,7 @@ def __init__(self,
                  td_error_loss_fn=element_wise_squared_loss,
                  td_lambda=0.95,
                  normalize_advantages=True,
+                 compute_advantages_internally=False,
```
doc string for this arg

breakds replied:
Done.

emailweixu commented on alf/algorithms/ppo_loss.py:
```diff
@@ -129,4 +131,6 @@ def _pg_loss(self, info, advantages):
-        return info.returns, info.advantages
+        if not self._compute_advantages_internally:
+            return info.returns, info.advantages
+        return super()._calc_returns_and_advantages(info, value)
```
should use the value from rollout to calculate the advantage and return

breakds replied:
Updated this (and also that in `ppg_aux_loss`) so that they use the value target based on the rollout time value prediction.

emailweixu commented on alf/algorithms/ppg_algorithm.py:
```diff
@@ -0,0 +1,468 @@
+        original_num_updates_per_train_iter = self._config.num_updates_per_train_iter
+
+        try:
+            self._aux_phase.on_context_switch_from(self._policy_phase)
```
Putting line 167 here and 358 directly can make the code a bit easier to read.

breakds replied:
Agree, updated.

breakds committed:
Add documentation to Disjoint Policy Value Network stating about the implementation differences

breakds committed:
Save rollout time value prediction and used that to compute GAE in training

breakds committed:
Address comments

breakds commented:
Updated by addressing the comments and integrated with #1036. PTAL.

Really appreciate your thorough review as well as the discussions @emailweixu . I did learn a lot for this!

emailweixu commented on alf/algorithms/ppo_loss.py:
```diff
@@ -133,4 +133,6 @@ def _pg_loss(self, info, advantages):
         if not self._compute_advantages_internally:
             return info.returns, info.advantages
-        return super()._calc_returns_and_advantages(info, value)
+        value_to_use = info.rollout_value if hasattr(
```
value_to_use = getattr(info, 'rollout_value', info.value)

emailweixu replied:
value_to_use = getattr(info, 'rollout_value', info.value)

emailweixu commented on alf/algorithms/ppg_algorithm.py:
```diff
@@ -0,0 +1,461 @@
+            self._replay_buffer = previous_replay_buffer
+            self._config.mini_batch_length = original_mini_batch_length
+            self._config.mini_batch_size = original_mini_batch_size
+            self._config.num_updates_per_train_iter = original_num_updates_per_train_iter
```
The code might be cleaner if aux-phase training is put in a sub-algorithm (say PPGAuxAlgorithm) so that we don't need to have this unnatural and potentially unsafe context switch (as it depends on too much detail of `train_from_replay_buffer()`) .
This PPGAuxAlgorithm will hold aux copy of the model, its own replay buffer and optimizer. We can all the PPGAuxAlgorithm.train_from_replay_buffer() to do the training without context switch. At PPGAlgorithm.rollout_step, we can call PPGAuxAlgorithm.observe_for_replay to store the experience.

emailweixu replied:
Another possible benefit is that we may be able to get rid of the problem of maintaining two copy of the model. This can be done by putting PPGAuxAlgorithm into _trainable_attributes_to_ignore of PPGAlgorithm

breakds replied:
Good point. I think I understand how this is supposed to work now. Let me give it a try.

breakds replied:
Algorithm is updated with a separate `PPGAuxAlgorithm`. Much cleaner!

breakds committed:
getattr handles default value

breakds committed:
Refactor to have a separate ppg_aux_algorithm

breakds committed:
Add documentation

breakds commented:
The PR is updated with a separate auxiliary update. PTAL.

Did another round of training to verify that it has the similar performance:

![ppg_new](https://user-images.githubusercontent.com/1111035/138610467-5c2e3a17-c7f4-4bb7-9cd5-7f2b2a8b2aa6.png)

In the above graph, 

1. green = the refactored implementation
2. light blue = the previous implementation before the refactor

The performance is very close.

emailweixu commented on alf/algorithms/ppg_algorithm.py:
```diff
@@ -0,0 +1,205 @@
+
+    def train_step(self, inputs: TimeStep, state,
+                   plain_rollout_info: PPGRolloutInfo) -> AlgStep:
+        """Phase context dependent evaluation on experiences for training
```
outdated comment

breakds replied:
Removed.

emailweixu commented on alf/algorithms/ppg_algorithm.py:
```diff
@@ -0,0 +1,205 @@
+        return alg_step._replace(info=train_info)
+
+    def calc_loss(self, info: PPGTrainInfo) -> LossInfo:
+        """Phase context dependent loss function evaluation
```
outdated comment

breakds replied:
Removed.

emailweixu commented on alf/algorithms/ppg_algorithm.py:
```diff
@@ -0,0 +1,205 @@
+        if alf.summary.get_global_counter(
+        ) % self._aux_algorithm.interval == 0:
+            # NOTE(breakds): The auxiliary steps ``aux_steps`` is
+            # currently not used.
```
so you don't need to assign to aux_steps

breakds replied:
Removed. Originally I was thinking the note and the assignment can be left here to remind the reader that `train_from_replay_buffer` does return the number of steps.

emailweixu commented on alf/algorithms/ppg/ppg_aux_algorithm.py:
```diff
@@ -0,0 +1,184 @@
+        """
+        # Use a config with overriden fields as auxiliary phase updates has a
+        # bunch of its own options.
+        updated_config = copy.deepcopy(config)
```
why not copy.copy()?

breakds replied:
I was being over-cautious :) Apparently if it is only updating POD fields we don't have to do `deepcopy`. Updated.

hnyu commented on alf/algorithms/ppg/ppg_aux_algorithm.py:
```diff
@@ -0,0 +1,184 @@
+
+        Args:
+
+            observation_spec (nested TensorSpec): representing the observations.
```
If type hints are already used, the types in docstrings can be omitted.

breakds replied:
Good point. I removed the type hints here because it should be nested TensorSpec. However, we do not have a type hint for nested `TensorSpec` yet.

hnyu commented on alf/algorithms/ppg/ppg_aux_algorithm.py:
```diff
@@ -0,0 +1,184 @@
+            exp = exp._replace(state=())
+        # Set the experience spec explicitly if it is not set, based on this
+        # (sample) experience
+        if not self._experience_spec:
```
Can this code block just reuse _set_exp_replayer defined in algorithm.py?

breakds replied:
Yes, but it does not do much to avoid code duplication - we still need to compute the replay buffer parameters explicitly. Besides, `_set_replay_buffer` does extra thing to append to `self._observers`, which does no harm but is not necessary.

hnyu commented on alf/algorithms/ppg/ppg_aux_algorithm.py:
```diff
@@ -0,0 +1,184 @@
+            exp_spec = dist_utils.to_distribution_param_spec(
+                self._experience_spec)
+            num_envs = exp.env_id.shape[0]
+            max_length = self._config.unroll_length
```
If PPGAuxOptions.interval > 1 and we only store unroll_length steps in the buffer, will some steps (within the interval) be discarded ?

breakds replied:
Here the `unroll_length` has an updated value because Aux phase use an updated `TrainerConfig`. Sorry that this is not as clear - I added a comment here to make it clearer.

hnyu commented on alf/algorithms/ppg/ppg_aux_phase_loss.py:
```diff
@@ -0,0 +1,135 @@
+                td_loss_aux=td_loss_aux,
+                policy_kl_loss=policy_kl_loss))
+
+    def _calc_returns_with_advantages(self, info, value):
```
The name is misleading. Can just be "_calc_returns" because you are not returning the advantages. Maybe can directly derive from ActorCriticLoss and save some code of this function?

breakds replied:
Updated the name. Probably not important, but one of the reasons to duplicate part of the code is to save the part where `returns` is calculated but not used for `gae` case.

hnyu commented on alf/algorithms/ppg_algorithm.py:
```diff
@@ -0,0 +1,205 @@
+        self._predict_step_epsilon_greedy = epsilon_greedy
+
+    def _trainable_attributes_to_ignore(self):
+        return ['PPGAuxAlgorithm']
```
Should this be the actual variable name "_aux_algorithm" instead of the class name?

breakds replied:
Sure. Updated the name to be `_aux_algorithm`.

hnyu replied:
> Sure. Updated the name to be `_aux_algorithm`.

Does this mean that you were not actually excluding the parameters from its parent?

breakds replied:
Did you mean `_trainable_attributes_to_ignore`? If so I have updated both places.

hnyu replied:
> Did you mean `_trainable_attributes_to_ignore`? If so I have updated both places.

Sorry I didn't explain it clearly. The reason why we have this function is to make sure that the parameters of a child won't be modified by its parent. So if originally the attribute names were wrong, does it mean that it's possible when you were training the algorithm, the child's parameters were modified by its parent in an unexpected way?

breakds replied:
Ah I think I misunderstood your comments previously. Let me explain:

1. My understanding is that here we use the `name` field of the `PPGAuxAlgorithm` instance to exclude it, instead of the variable name.
2. Originally the `name` field is forced to set to `PPGAuxAlgorithm`, which is also the class name. And therefore originally I was excluding it using `PPGAuxAlgorithm`.
3. I mistook your comments as "it is better to name it `_aux_algorithm` to avoid being identical to the class name". Therefore, I have updated both places to be `_aux_algorithm`, which I think it should still work.

hnyu replied:
From what I know, for this function we need to provide the variable name. For example, if you have a class member called `self._vars` then the list should be ["_vars"]. It has nothing to do with the name you assign to that class instance. See the example of SAC's "_target_critic_networks". In other words, it's possible that you weren't actually excluded its parameters from the parent, because the wrong name was provided.

breakds replied:
Ah Yes you are right. I should have carefully read the documentation. Thanks for pointing this out.

And yes, previously it was not correctly ignored.

emailweixu replied:
Since your result is based on using "PPGAuxAlgorithm", which does not have any effect, how does it affect the result?

hnyu commented on alf/algorithms/ppo_loss.py:
```diff
@@ -68,6 +68,11 @@ def __init__(self,
+                compute the adavantage and it expects the info to carry the
+                already-computed advantage. If this flag is set to True, PPOLoss
+                will instead compute the advantage internally without depending
+                on the input info.
```
Should additionally explain in what scenarios we need to compute it internally because normally PPO's advantages stay fixed during the updates, so we tend to only compute once before training.

breakds replied:
Good point. The direct reason is that we want to avoid having `preprocess_experience()`, but I am not sure this still holds after splitting the algorithms. Let me think and confirm.

breakds replied:
Added documentation that explains the reason in the docstring.

hnyu commented on alf/examples/ppg_conf.py:
```diff
@@ -0,0 +1,32 @@
+from alf.algorithms.agent import Agent
+from alf.algorithms.ppg_algorithm import PPGAlgorithm
+from alf.algorithms.ppo_algorithm import PPOLoss
+
```
Can you also change the ppo_conf.py so that by default the enforce_entropy_target=False?

breakds replied:
Will do in a separate PR. Sorry it seems that I forgot about this previously.

emailweixu commented on alf/algorithms/ppg/ppg_aux_algorithm.py:
```diff
@@ -0,0 +1,184 @@
+
+    def train_step(self, inputs: TimeStep, state,
+                   plain_rollout_info: PPGRolloutInfo) -> AlgStep:
+        """Phase context dependent evaluation on experiences for training
```
no more phase context

breakds replied:
Removed.

emailweixu commented on alf/algorithms/ppg/ppg_aux_algorithm.py:
```diff
@@ -0,0 +1,184 @@
+        return alg_step._replace(info=train_info)
+
+    def calc_loss(self, info: PPGTrainInfo) -> LossInfo:
+        """Phase context dependent loss function evaluation
```
no more phase context

breakds replied:
Removed.

emailweixu commented on alf/algorithms/ppg/ppg_utils.py:
```diff
@@ -0,0 +1,140 @@
+        PPGRolloutInfo) from evaluation of the updated network. This function
+        makes it easy to do that.
+
+        It is also used in ``preprocess_experience()`` where we need to combine
```
there is no preprocess_experience

emailweixu commented on alf/algorithms/ppg/ppg_utils.py:
```diff
@@ -0,0 +1,140 @@
+    in addition to the original rollout information.
+
+    Such extra information is derived during training updates, in
+    ``preprocess_experience()`` and used across calls to
```
there is no preprocess_experience

breakds replied:
Good catch. Removed.

breakds committed:
Addressed comments

breakds committed:
Use copy instead of deepcopy

breakds committed:
There is no preprocess_experience() any more.

emailweixu commented on alf/algorithms/ppg/ppg_aux_algorithm.py:
```diff
@@ -0,0 +1,183 @@
+        updated_config.mini_batch_length = (aux_options.mini_batch_length
+                                            or config.unroll_length)
+        updated_config.mini_batch_size = aux_options.mini_batch_size
+        updated_config.num_updates_per_train_iter = aux_options.num_updates_per_train_iter
```
should clear data_transformer too. data_transformer is used to process the untransformed data. The data stored in the replay buffer of PPGAuxAlgorithm is already processed by Agent.

breakds replied:
Right. Fixed by setting it to `None` and adding documentation explaining why.

breakds committed:
Add documentation explaining why compute_advantages_internally is needed in PPOLoss

breakds committed:
No need to explicitly assign a name to PPGAuxAlgorithm instance

breakds committed:
Explicitly set data transformer to None for PPGAuxAlgorithm

breakds commented:
> Since your result is based on using "PPGAuxAlgorithm", which does not have any effect, how does it affect the result?

Not sure yet. Running the experiment to confirm this.

breakds commented:
> > Since your result is based on using "PPGAuxAlgorithm", which does not have any effect, how does it affect the result?
> 
> Not sure yet. Running the experiment to confirm this.

The experiment shows the new refactor results in the same training dynamics and performance (i.e. previously the bug in `_trainnable_attributes_to_ignore` does not make a difference in the current experiment setting for procgen bossfight).

![updated_ppg](https://user-images.githubusercontent.com/1111035/139140167-ee90c867-95f6-4623-ad95-0aa3152b3588.png)

hnyu commented on alf/algorithms/ppo_loss.py:
```diff
@@ -129,4 +138,12 @@ def _pg_loss(self, info, advantages):
+        # info.value is the newly computed value which is different from the
+        # rollout time value prediction. The rollout time value prediction is
+        # preserved in info.rollout_value.
+        value_to_use = getattr(info, 'rollout_value', info.value)
```
The default value "info.value" seems problematic if 'rollout_value' doesn't exist? I suggest just asserting to False

breakds replied:
Sure. Updated this with an assert to make it more explicit to the user/reader.

breakds committed:
Explicitly assert rollout_value to be present when _compute_advantages_internally
